### PR TITLE
[SPARK-7699][Core] Lazy start the scheduler for dynamic allocation

### PR DESCRIPTION
**Pull request by jerryshao**
This patch propose to lazy start the scheduler for dynamic allocation to avoid fast ramp down executor numbers is load is less.

This implementation will:
1. immediately start the scheduler is `numExecutorsTarget` is 0, this is the expected behavior.
2. if `numExecutorsTarget` is not zero, start the scheduler until the number is satisfied, if the load is less, this initial started executors will last for at least 60 seconds, user will have a window to submit a job, no need to revamp the executors.
3. if `numExecutorsTarget` is not satisfied until the timeout, this means resource is not enough, the scheduler will start until this timeout, will not wait infinitely.

Please help to review, thanks a lot.

### CODE REVIEW

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by sryza** [comment]
If my understanding is correct, this affects the add behavior, making it so that `targetNumExecutors` can't increase until 60 seconds have passed or we've received all the initial requested executors.

Can we change the logic to be that we don't reduce `targetNumExecutors` below initialExecutors until either
* a stage has been submitted
* we've waited for `spark.dynamicAllocation.executorIdleTimeout` seconds

**Comment by jerryshao** [comment]
Hey @sryza , so you mean we only consider the `targetNumExecutors` below than `initialExecutors` situation, if the load is heavy, immediately increase the `targetNumExecutors`, is that right ?

**Comment by sryza** [comment]
exactly

**Comment by jerryshao** [comment]
OK, let me change the logic to handle this.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -240,6 +244,7 @@ private[spark] class ExecutorAllocationManager(
     removeTimes.retain { case (executorId, expireTime) =>
       val expired = now >= expireTime
       if (expired) {
+        numTargetExecutorAdjustable.compareAndSet(false, true)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: These `compareAndSet` can just be `set` I think. In fact a `@volatile boolean` is OK too. I don't know what's simpler.

**Comment by srowen** [comment]
This looks like a pretty reasonable change, to attempt to defer evaluating a change in executors until something happens. I suppose this still means that something like `spark-shell` will never ramp down until you execute a job, but, maybe that's fine?

**Comment by jerryshao** [comment]
`spark-shell` will also ramp down until the the first job is submitted or the initial executor is idle timeout, no need to wait until the first job is submitted, by default after 60 seconds `numTargetExecutor` can  be re-calculated.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 1 above)
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: +1 to using a @volatile boolean

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 2]
```diff
@@ -150,6 +151,9 @@ private[spark] class ExecutorAllocationManager(
   // Metric source for ExecutorAllocationManager to expose internal status to MetricsSystem.
   val executorAllocationManagerSource = new ExecutorAllocationManagerSource
 
+  // Flag to measure whether numExecutorTarget could be adjusted
+  val numTargetExecutorAdjustable = new AtomicBoolean(false)
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Can we rename this to `inInitialExecutorsPeriod` and add a comment explaining the criteria for being in this period?

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 3]
```diff
@@ -262,15 +267,22 @@ private[spark] class ExecutorAllocationManager(
...
-      // executors and inform the cluster manager to cancel the extra pending requests
-      val oldNumExecutorsTarget = numExecutorsTarget
-      numExecutorsTarget = math.max(maxNeeded, minNumExecutors)
-      client.requestTotalExecutors(numExecutorsTarget)
-      numExecutorsToAdd = 1
-      logInfo(s"Lowering target number of executors to $numExecutorsTarget because " +
-        s"not all requests are actually needed (previously $oldNumExecutorsTarget)")
-      numExecutorsTarget - oldNumExecutorsTarget
+      if (!numTargetExecutorAdjustable.get) {
+        // Keep the initial number of target executor to not ramp down until the first job is
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: IIUC, this will stop us from adjusting our target down to something greater than initialExecutors.  E.g. if initialExecutors is 5 and we raise targetNumExecutors to 10, we won't then be able to decrease it down to 7.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Why? This is only in the initial period, where we shouldn't even be adjusting our target in the first place.

**Comment by andrewor14** [comment]
Hi all, I looked at the code and I think it works. However, I think there is a slightly simpler alternative.
```
// in updateAndSyncNumExecutorsTarget
val maxNeeded =
  if (initializing) {
    initialExecutors
  } else {
    maxNumExecutorsNeeded 
  }
```
Then the rest of the code in `updateAndSyncNumExecutorsTarget` can remain the same.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I would just call this `initializing`

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Then could you add a comment to explain what this means, maybe something like:
```
// Whether we are still waiting for the initial set of executors to be allocated.
// If this is true, we will not cancel outstanding executor requests. This is set
// to false when (1) a stage is submitted, or (2) an executor idle timeout has elapsed.
```

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala [hunk 4]
```diff
@@ -697,13 +697,65 @@ class ExecutorAllocationManagerSuite extends FunSuite with LocalSparkContext wit
...
+    assert(numExecutorsTarget(manager) === 3)
+
+    // Schedule again to recalculate the numExecutorsTarget after executor is timeout
+    schedule(manager)
+    // Verify that current number of executors should be ramp down when executor is timeout
+    assert(numExecutorsTarget(manager) === 2)
+  }
+
+  private def createSparkContext(minExecutors: Int = 1, maxExecutors: Int = 5,
+      initialExecutors: Int = 1): SparkContext = {
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: style:
```
private def createSparkContext(
    minExecutors: Int = 1,
    maxExecutors: Int = 5,
    initialExecutors: Int = 1): SparkContext = {
  ...
}
```

Comment (sryza) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: @andrewor14 it's the other way around.  During the initialization period, we never want to go below initialExecutors, but we're ok going above it.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 3 above)
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: That's not that approach that I thought we settled on.  Is there a basis for that which I'm missing?

**Comment by sryza** [comment]
@andrewor14  it's the other way around.  We're ok with going above initialExecutors during the initialization period, but don't want to go below it.

**Comment by andrewor14** [comment]
Hm, that seems to be so. I'm wondering if there's a better way to simplify the cases so we don't have nested ifs. Previously it looks like this:
```
if (needed < target) {
  // cancel
} else if (need more executors) {
  // add
}
```
Now we add some new initialization logic in the cancel case. What if we express it like:
```
if (initializing && numExecutors != initialExecutors) {
  // resync with cluster manager
} else if ... // as before
```
It's fairly minor but I think it's simpler to reason about.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 3 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: is it even possible to adjust it downwards during the initial window though? There are two ways of doing that, through cancel or remove. Here we can't cancel, and by the time remove happens we're no longer in the initialization window anymore.

**Comment by andrewor14** [comment]
Another related problem with or without my suggested change: we actually call `updateAndSync...` quite often (every 100ms). While we're initializing we'll end up flooding the AM with `requestTotalExecutors` messages. Maybe this is not a huge problem but it's something to look out for.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 5]
```diff
@@ -262,15 +267,22 @@ private[spark] class ExecutorAllocationManager(
...
+        // submitted or the first idle executor is released.
+        client.requestTotalExecutors(numExecutorsTarget)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Actually, does it make sense to do this inside the case `maxNeeded < numExecutorsTarget`? This condition implies that we are targeting more executors that are needed, and so we should cancel outstanding requests. This is not really true during initialization.

I think a better indication that your application has initialized is if the number of executors is >= your initial executors.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 3 above)
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: What do you mean by "we can't cancel?"  Why not?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: because `numTargetExecutorAdjustable` is false, right?

Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: That's what I'm saying we should change about this code. (Sorry if I haven't been communicating this well or am missing something obvious).

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Wait, then I'm pretty confused. Once we set `numTargetExecutorAdjustable` to true, we *can* ramp it down below the initial number of executors. It's just that while we're still initializing this can't happen (which is the whole point of this patch). Is that not the intended behavior?

Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: That is the intended behavior.  My complaint is that this patch makes it so that, in addition to that behavior, during the initialization period we can't ramp our target down at all, even if the new target would stay above the initial number of executors.  Does that make sense?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Yeah, now I get it.

But during the initialization phase, we can't actually ramp up beyond initial executors can we? This is because the only way to ramp up is to have a backlog of tasks, and this only happens after we have submitted a stage, at which point we are no longer in the initialization phase.

Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Ah, you're totally right.  So we should just add comments that make this clear.

**Comment by sryza** [comment]
To avoid the problem @andrewor14 brought up, can we avoid calling `requestTotalExecutors` when numExecutorsTarget hasn't changed.?

**Comment by jerryshao** [comment]
Thanks a lot @andrewor14 and @sryza for your comments, sorry for late response, I will try to understand your comments and change the code accordingly :).

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 2 above)
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Hi @andrewor14 , do we need to consider this condition `(1) the initial executors have all registered` to change the `initializing` to false?

Previously in this patch we set `initializing` to false  when: 
1. a stage is submitted.
2. an executor is timeout due to idle (default is 60 secs)

Now we add another condition, so possibly we may have such scenario: when all the initial executors are registered, but no stage is submitted and executor is not idle timeout. So the executors can have a chance to ramp down at this time window, I think this will introduce additional requesting when stage is later submitted, is this a intended behavior?

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 6]
```diff
@@ -262,15 +267,22 @@ private[spark] class ExecutorAllocationManager(
...
+        0
+      } else {
+        // The target number exceeds the number we actually need, so stop adding new
+        // executors and inform the cluster manager to cancel the extra pending requests
+        val oldNumExecutorsTarget = numExecutorsTarget
+        numExecutorsTarget = math.max(maxNeeded, minNumExecutors)
+        client.requestTotalExecutors(numExecutorsTarget)
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: @sryza and @andrewor14 , do we need to avoid requesting executors also here and `addExecutors` when `oldNumExecutorsTarget` == `numExecutorsTarget`, not in initializing status?

**Comment by jerryshao** [comment]
Hi all, I just updated the code according to your comments, please help to review, thanks a lot. There's two opening questions I didn't address yet:

1. Do we need to set the `initializing` flag to false when the initial executors have all registered. There's a concern as I mentioned before, there's a window for Yarn to ramp down the executors when all the initial executors are registered, but user job has not yet submitted and executor is not timeout, I'm not sure is this a intended behavior?
2. Do we need to change the code to not send request when the old `numExecutorsTarget` is equal to new calculated one in all the situations (when initializing is finished)?

Please help to address the concern, thanks a lot and appreciate your time.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 7]
```diff
@@ -261,14 +270,19 @@ private[spark] class ExecutorAllocationManager(
...
+      0
+    } else if (!initializing && maxNeeded < numExecutorsTarget) {
       // The target number exceeds the number we actually need, so stop adding new
       // executors and inform the cluster manager to cancel the extra pending requests
       val oldNumExecutorsTarget = numExecutorsTarget
       numExecutorsTarget = math.max(maxNeeded, minNumExecutors)
       client.requestTotalExecutors(numExecutorsTarget)
       numExecutorsToAdd = 1
-      logInfo(s"Lowering target number of executors to $numExecutorsTarget because " +
+      logDebug(s"Lowering target number of executors to $numExecutorsTarget because " +
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: This is a little verbose, so change to `logDebug`.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 8]
```diff
@@ -261,14 +270,19 @@ private[spark] class ExecutorAllocationManager(
   private def updateAndSyncNumExecutorsTarget(now: Long): Int = synchronized {
     val maxNeeded = maxNumExecutorsNeeded
 
-    if (maxNeeded < numExecutorsTarget) {
+    if (initializing && (maxNeeded < numExecutorsTarget)) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Bring the nested structure out, not sure is there any better and simple way.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 7 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: ok, but this conflicts with a patch I just merged

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 6 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: this was fixed in 51898b5158ac7e7e67b0539bc062c9c16ce9a7ce

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 8 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: the second part doesn't seem necessary. I don't think `maxNeeded` can be >= `numExecutorsTarget` if we are still initializing, since we only increase `maxNeeded` after submitting tasks.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 9]
```diff
@@ -261,14 +269,19 @@ private[spark] class ExecutorAllocationManager(
...
+      // when ExecutorAllocationManager is still in initializing status,
+      // and the max number of needed executors is less than the initial number of executors,
+      // we just return without doing anything.
+      0
+    } else if (!initializing && maxNeeded < numExecutorsTarget) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: This is pretty long-winded. I think we can just do
```
if (initializing) {
} else if (maxNeeded < numExecutorsTarget) {
} ...
```

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 10]
```diff
@@ -261,14 +269,19 @@ private[spark] class ExecutorAllocationManager(
   private def updateAndSyncNumExecutorsTarget(now: Long): Int = synchronized {
     val maxNeeded = maxNumExecutorsNeeded
 
-    if (maxNeeded < numExecutorsTarget) {
+    if (initializing && (maxNeeded < numExecutorsTarget)) {
+      // when ExecutorAllocationManager is still in initializing status,
+      // and the max number of needed executors is less than the initial number of executors,
+      // we just return without doing anything.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I would rephrase this:
```
// Do not cancel requests while we are still initializing,
// otherwise the first job may have to ramp up unnecessarily
```

**Comment by andrewor14** [comment]
@jerryshao Thanks for the update. The latest changes look fairly reasonable.

(1) I think you're correct. We don't actually need to wait for the initial set of executors to register to set `initializing = false`. Before the first stage is submitted, we can't actually ever ramp up our target. We can't cancel either, however, because all executors have already registered. In other words, setting `initializing = false` in this case does nothing.

(2) I was thinking we don't actually need to sync with the cluster manager to get our executors while we're initializing. We don't do that currently, and I don't see why we need to add that. This `initializing` flag is just a way for us to delay canceling. Then we don't need to worry about spamming the AM with sync messages while we initialize.

**Comment by jerryshao** [comment]
Hi @andrewor14 , I just updated the code according to your comments, please help to review, thanks a lot.

**Comment by andrewor14** [comment]
lgtm. @jerryshao have you had a chance to test this on a real cluster?

**Comment by jerryshao** [comment]
I just tried in my local pseudo-cluster for functionality, haven't tried in real cluster for performance, I will test it.

**Comment by andrewor14** [comment]
@sryza I'm planning to pull this in. Any other comments?

**Comment by sryza** [comment]
This looks right to me.  Merging.
